EC2m LD ladder

romen opened the pull request:
# TL;DR

PR #6009 and #6070 unified the code path for EC point multiplication for prime
and binary curves, introducing several SCA defenses.
For binary curves the impact on performance of the security improvements was
very high (as reported in some GitHub issues, e.g., #6396 ).

This PR aims at improving the existing `ec_mul_consttime` function introducing
hooks that an `EC_METHOD` could use to provide specialized, and faster,
implementations.

We then provide such a specialized implementation that applies to binary curves,
bringing the performance of cryptosystems built on top of these reasonably close
to the performance level of `1.1.1-pre3` (before most of the new SCA
countermeasures were applied).

In most instances (see the detailed table below) the new implementation is more
than 10x times faster than current master.

The new implementation is still (reasonably) slower than `1.1.1-pre3`, but:
-  latest master includes other changes to the GF2 arithmetic not immediately
   related to this PR (e.g., 0dae8baf , b336ce57 , ... );
-  latest master includes also other SCA countermeasures which have an impact on
   performance (see, e.g., #6664 );
-  `1.1.1-pre3` is leaky in, e.g., bit length for keygen and derive;
-  `1.1.1-pre3` has no bug attack or DPA mitigations.

This PR only affects binary curves, but the scaffold introduced by the first
commit opens the way for improving in a similar way also prime curves, which
also suffered a slowdown as a consequence of the recent SCA hardening process,
although minor in comparison to the effect on binary curves.

Co-authored by @sohhas and @bbbrumley .

# Benchmark table

The following table presents the CPU cycles average computed over 1000 runs per
operation.

*The benchmarks were executed on Intel Core i7-6700 CPU, running at 3.40GHz, with
Turbo Boost disabled, no frequency scaling and on a single core reserved only
for the execution of the benchmark, while the rest of the system was mostly
idle.*

| Curve/Operation  |    openssl-111-pre3   | openssl-111-pre9-dev | openssl-ec2m_ladder  |   ec2m_ladder/pre3   | ec2m_ladder/pre9-dev |
| ---------------- | --------------------: | -------------------: | -------------------: | -------------------: | -------------------: |
|  **sect163r2**   |                       |                      |                      |                      |                      |
| evp_pkey_keygen  |               675 105 |            7 415 437 |              809 969 |                 1.20 |                 0.11 |
| evp_pkey_derive  |               647 086 |            7 396 699 |              791 018 |                 1.22 |                 0.11 |
| evp_digestsign   |               716 298 |            7 586 516 |              914 920 |                 1.28 |                 0.12 |
| evp_digestverify |             1 331 206 |            5 370 424 |            1 638 221 |                 1.23 |                 0.31 |
|  **sect233r1**   |                       |                      |                      |                      |                      |
| evp_pkey_keygen  |               863 484 |           14 307 333 |              980 921 |                 1.14 |                 0.07 |
| evp_pkey_derive  |               835 200 |           14 266 140 |              973 475 |                 1.17 |                 0.07 |
| evp_digestsign   |               944 166 |           14 435 245 |            1 155 727 |                 1.22 |                 0.08 |
| evp_digestverify |             1 736 499 |           10 185 112 |            2 016 280 |                 1.16 |                 0.20 |
|  **sect283r1**   |                       |                      |                      |                      |                      |
| evp_pkey_keygen  |             1 513 133 |           22 023 374 |            1 742 583 |                 1.15 |                 0.08 |
| evp_pkey_derive  |             1 479 988 |           21 978 168 |            1 732 546 |                 1.17 |                 0.08 |
| evp_digestsign   |             1 625 851 |           22 238 791 |            1 972 981 |                 1.21 |                 0.09 |
| evp_digestverify |             3 054 693 |           15 508 890 |            3 592 322 |                 1.18 |                 0.23 |
|  **sect571r1**   |                       |                      |                      |                      |                      |
| evp_pkey_keygen  |             5 601 491 |          106 207 308 |            6 328 737 |                 1.13 |                 0.06 |
| evp_pkey_derive  |             5 525 044 |          106 119 035 |            6 320 248 |                 1.14 |                 0.06 |
| evp_digestsign   |             6 014 458 |          106 691 078 |            7 118 188 |                 1.18 |                 0.07 |
| evp_digestverify |            11 368 181 |           73 156 913 |           13 180 169 |                 1.16 |                 0.18 |


# Commits breakout

## EC point multiplication: add ladder scaffold for specialized Montgomery ladder implementations

PR #6009 and #6070 replaced the default EC point multiplication path for
prime and binary curves with a unified Montgomery ladder implementation
with various timing attack defenses (for the common paths when a secret
scalar is feed to the point multiplication).
The newly introduced default implementation directly used
EC_POINT_add/dbl in the main loop.

The scaffolding introduced by this commit allows EC_METHODs to define a
specialized `ladder_step` function to improve performances by taking
advantage of efficient formulas for differential addition-and-doubling
and different coordinate systems.

- `ladder_pre` is executed before the main loop of the ladder: by
default it copies the input point P into S, and doubles it into R.
Specialized implementations could, e.g., use this hook to transition
to different coordinate systems before copying and doubling;
- `ladder_step` is the core of the Montgomery ladder loop: by default it
computes `S := R+S; R := 2R;`, but specific implementations could,
e.g., implement a more efficient formula for differential
addition-and-doubling;
- `ladder_post` is executed after the Montgomery ladder loop: by default
it's a noop, but specialized implementations could, e.g., use this
hook to transition back from the coordinate system used for optimizing
the differential addition-and-doubling or recover the y coordinate of
the result point.

**Note:** requires a `make update` for being atomic

## [test] test some important ladder corner cases and catch corner cases better and earlier

*Self explicative, you might consider squashing this commit with the previous
one; I decided to keep it separate as it is technically independent from any
other change in this PR and because it was authored by @bbbrumley alone.*

## EC2M ladder: Lopez-Dahab ladder implementation

This commit uses the new ladder scaffold to implement a specialized
ladder step based on differential addition-and-doubling in mixed
Lopez-Dahab projective coordinates, modified to independently blind the
operands.

The arithmetic in `ladder_pre`, `ladder_step` and `ladder_post` is
auto generated with tooling:
- see, e.g., "Guide to ECC" Alg 3.40 for reference about the
  `ladder_pre` implementation;
- see https://www.hyperelliptic.org/EFD/g12o/auto-code/shortw/xz/ladder/mladd-2003-s.op3
  for the differential addition-and-doubling formulas implemented in
  `ladder_step`;
- see, e.g., "Fast Multiplication on Elliptic Curves over GF(2**m)
  without Precomputation" (Lopez and Dahab, CHES 1999) Appendix Alg Mxy
  for the `ladder_post` implementation to recover the `(x,y)` result in
  affine coordinates.

**Note:** requires a `make update` for being atomic

## EC2M LD ladder: use it also for ECDSA verify

By default `ec_mul_consttime` (which uses the ladder implementation) is
used only for k*Generator or k*GenericPoint.
ECDSA verification uses a*Generator+b*GenericPoint: this commit
forces the use of `ec_mul_consttime` (and the ladder) also for the ECDSA
verification path, while using the default wNAF implementation for any
other case.

*With this commit `ec_mul_consttime` loses the static attribute, and is added to
`ec_lcl.h` so `EC_METHOD`s can directly use it.*

**Note:** requires a `make update` for being atomic

*Again this is technically independent from the parent commit, but you might
consider squashing it with the parent: without this ECDSA verification using
binary curves does not benefit from the performance boost that the new
implementation brings*

## fixme: make update

*Given that I don't know how you would prefer to reorganize the commits, rather
than running `make update` for each commit above, I made a single `make update`
commit for the whole PR: depending on your feedback I will interactively rebase
this PR to split the changes introduced in the error codes so that each final
commit is atomic and leads to a successful compilation.*

# Checklist
- [x] documentation is added or updated
- [x] tests are added or updated
- [ ] make commits atomic (running `make update`) once we receive feedback

romen commented on crypto/ec/ec_err.c:
```diff
@@ -261,12 +268,6 @@ static const ERR_STRING_DATA EC_str_functs[] = {
-    {ERR_PACK(ERR_LIB_EC, EC_F_PKEY_SM2_INIT, 0), "pkey_sm2_init"},
-    {ERR_PACK(ERR_LIB_EC, EC_F_PKEY_SM2_KEYGEN, 0), "pkey_sm2_keygen"},
-    {ERR_PACK(ERR_LIB_EC, EC_F_PKEY_SM2_PARAMGEN, 0), "pkey_sm2_paramgen"},
-    {ERR_PACK(ERR_LIB_EC, EC_F_PKEY_SM2_SIGN, 0), "pkey_sm2_sign"},
```
I'm not sure why `make update` wants to remove SM2 stuff, should this have been removed in a previous PR?

romen replied:
@richsalz am I doing something wrong here? Should I use something more specific to update the error codes, rather than a full `make update` to avoid removing this SM2 stuff?

richsalz replied:
I'm not sure what's happening.  Maybe you're not on the current master?  Curious.

romen replied:
Fixed following @dot-asm advice!

romen commented:
I just rebased on master after improving the "EC2M LD ladder: use it also for ECDSA verify" commit, hardening it against weird arguments that could be passed in but are not currently covered by testing.

I actually moved most of those checks to the `EC_POINTs_mul` wrapper, as they are actually independent from the specific implementation and should always be executed.

/cc @bbbrumley (that gave me the original input for the latest changes) for feedback on 063e3d9

romen commented:
Ping @richsalz for an opinion about moving the checks to the wrapper (and if that should actually be ported to 1.1.0 as well) and possibly also on how OMC would prefer the commits to be organized.

I know @mattcaswell is on holiday, does anyone else from the committers care to give feedback on this? Maybe @dot-asm ?

richsalz commented:
I'm not worried about how the commits are organized, although others might be.  Moving the checks to the wrapper, and out of the internal functions is reasonable if those aren't part of the public API.

richsalz commented:
@davidben or @dot-asm any feedback?

dot-asm commented on crypto/ec/ec_lib.c:
```diff
@@ -919,11 +919,39 @@ int EC_POINTs_mul(const EC_GROUP *group, EC_POINT *r, const BIGNUM *scalar,
+        ret = ec_wNAF_mul(group, r, scalar, num, points, scalars, ctx);
 
-    return group->meth->mul(group, r, scalar, num, points, scalars, ctx);
+    if (new_ctx)
```
Explicit new_ctx != NULL is preferred nowadays.

dot-asm commented on crypto/ec/ecp_nistp224.c:
```diff
@@ -1410,9 +1413,6 @@ int ec_GFp_nistp224_points_mul(const EC_GROUP *group, EC_POINT *r,
 
-    if (ctx == NULL)
-        if ((ctx = new_ctx = BN_CTX_new()) == NULL)
-            return 0;
```
In such case you can remove new_ctx altogether. This comment applies even below.

dot-asm commented on crypto/ec/ec2_smpl.c:
```diff
@@ -737,4 +680,280 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+ err:
+    EC_POINT_free(q);
+    if (t != NULL)
+        EC_POINT_free(t);
```
EC_POINT_free checks for NULL, no need to do it here.

dot-asm commented:
The fact that some of errors are removed is no mystery. They are actually unreferenced, and should be removed. Note that they start with EC_F_PKEY, while all actual codes are SM2_F_PKEY. So right thing to do would be to run `make update [ERROR_REBUILD=-rebuild]` *prior* suggested modifications. "Right thing" in sense that there would be separate commit that removes unreferenced values, so that every commit would be about one thing, not two. It can be done in two ways. I can create separate request that removes them, and you can re-base later. Or you can re-master this request by injecting separate commit that removes the values in question.

romen commented:
@dot-asm I saw your commit only now, after replying to @richsalz above with the same answer.

> It can be done in two ways. I can create separate request that removes them, and you can re-base later. Or you can re-master this request by injecting separate commit that removes the values in question.

I will do a rebase now, with a commit to remove the old unused ones, followed by our new commits, making them "atomic" so that each of them will include only its own `make update`.

Thanks!

(I'll now temporarily mark this PR as WIP and ping you again when it's ready for review!)

dot-asm commented on crypto/ec/ec_lcl.h:
```diff
@@ -637,4 +646,73 @@ int X25519(uint8_t out_shared_key[32], const uint8_t private_key[32],
+
+static inline int ec_point_ladder_step(const EC_GROUP *group,
+                                      EC_POINT *r, EC_POINT *s,
+                                      EC_POINT *p, BN_CTX *ctx)
```
Formatting nit, two lines are one space too short.

romen replied:
thanks for catching this! I will fix both of them during the rebase!

dot-asm commented on crypto/ec/ec_lcl.h:
```diff
@@ -637,4 +646,73 @@ int X25519(uint8_t out_shared_key[32], const uint8_t private_key[32],
+
+static inline int ec_point_ladder_post(const EC_GROUP *group,
+                                      EC_POINT *r, EC_POINT *s,
+                                      EC_POINT *p, BN_CTX *ctx)
```
Formatting nit. Two lines are one space too short.

dot-asm commented on crypto/ec/ec_lcl.h:
```diff
@@ -178,6 +178,15 @@ struct ec_method_st {
     int (*blind_coordinates)(const EC_GROUP *group, EC_POINT *p, BN_CTX *ctx);
+    int (*ladder_pre)(const EC_GROUP *group,
+                       EC_POINT *r, EC_POINT *s,
+                       EC_POINT *p, BN_CTX *ctx);
```
Formatting nit, two lines are one space too long.

romen replied:
Fixed!

dot-asm commented on crypto/ec/ec2_smpl.c:
```diff
@@ -737,4 +680,279 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+}
+
+/*-
+ * Ladder step: differential addition-and-doubling, mixed LD coords.
```
Even though it's kind of obvious from the context I'd like to see "LD" expanded. Even elsewhere.

romen replied:
I fixed this in the code, it might be useful to rewrite the message for commit 84e3bdd722cc0, e.g.:

```
EC2M Lopez-Dahab ladder: use it also for ECDSA verify

By default `ec_mul_consttime` (which uses the ladder implementation) is
used only for k*Generator or k*GenericPoint.
ECDSA verification uses a*Generator+b*GenericPoint: this commit
forces the use of `ec_mul_consttime` (and the Lopez-Dahab ladder) also
for the ECDSA verification path, while using the default wNAF
implementation for any other case.

With this commit ec_mul_consttime loses the static attribute, and is
added to ec_lcl.h so EC_METHODs can directly use it.

While working on a new custom EC_POINTs_mul implementation, I realized
that many checks (e.g. all the points being compatible with the given
group, creating ctx if NULL, scalar==NULL && num==0) were duplicated
again and again in every single implementation (and actually some
implementations lacked some tests). I thought that it makes way more
sense for those checks that are independent from the actual
implementation and should always be done, to be moved in the
EC_POINTs_mul wrapper: so this commit also includes these changes.
```

mattcaswell commented on crypto/ec/ec2_smpl.c:
```diff
@@ -737,4 +680,279 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+            goto err;
+    }
+
+    if(!EC_POINT_copy(r,q))
```
Style nit: space needed after ","

mattcaswell commented on crypto/ec/ec_mult.c:
```diff
@@ -108,61 +108,76 @@ void EC_ec_pre_comp_free(EC_PRE_COMP *pre)
+    }
+
+    if (!BN_mul(cardinality, group->order, group->cofactor, ctx))
         goto err;
```
There seems to be inconsistency in this function about whether we call `ECerr` or not.

romen replied:
The "fixup!" commits below fix the reported formatting nits in the corresponding commits.

I will have now a look at `ec_mul_consttime` to raise `ECerr` in the appropriate cases.

romen replied:
@mattcaswell can you check 6f9fd585222 ?

(I couldn't find more specific reasons than `ERR_R_EC_LIB`/`ERR_R_BN_LIB`). Suggestions?

mattcaswell commented on test/ectest.c:
```diff
@@ -62,6 +70,8 @@ static int group_order_tests(EC_GROUP *group)
              */
             || !TEST_true(EC_POINT_mul(group, P, n1, NULL, NULL, ctx))
+            || (i == 1 && !TEST_int_eq(0, EC_POINT_cmp(group, P,
+                EC_GROUP_get0_generator(group), ctx)))
```
Indentation seems strange here.

dot-asm commented on crypto/ec/ec2_smpl.c:
```diff
@@ -737,4 +680,279 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+            goto err;
+
+        if (p != q && !EC_POINT_add(group, q, q, p, ctx))
+            goto err;
```
On additional thought, it's probably names being so generic that tips me off. For example if q was called say acc[umulator], it wouldn't be a problem. Point is that generic names make you wonder if there is any meaning behind letter. For example you would think of p as "product", and what would be q? And what is connection? With name like acc[umulator] you see directly that it's just a placeholder...

romen replied:
bbd0e2b is my attempt at maximizing readability: I clearly and explicitly separated the 3 cases: this comes at the "cost" of duplicating one line of code (`ec_mul_consttime(group, r, scalars[0], points[0], ctx)`) but I think it is a good tradeoff which does not uglify the code and solves way more problems in readability and upkeep than it creates.

dot-asm replied:
Cool!

dot-asm commented on crypto/ec/ec_lcl.h:
```diff
@@ -637,4 +646,73 @@ int X25519(uint8_t out_shared_key[32], const uint8_t private_key[32],
+ */
+int ec_mul_consttime(const EC_GROUP *group, EC_POINT *r,
+                     const BIGNUM *scalar, const EC_POINT *point,
+                     BN_CTX *ctx);
```
I for one sense inconsistency in this function naming. Now when it's shared, a better name might be more appropriate. Maybe ec_mul_by_scalar_consttime? Just me? Obviously side note...

romen replied:
What about `ec_scalarmul_ladder`?

I feel a bit of resistance in naming anything "consttime" given that we cannot promise anything about the underlying layer.

romen replied:
Once we agree on the renaming, given that it would require fixing the`make update` step and rewrite some commit messages, I would do a proper rebase, also squashing the stack of small fixup commits that piled up till now.

But I will have to wait until tomorrow to act upon this. 

I totally favor a renaming, for the reason above, let's just agree on a proper name. 

@bbbrumley/@mattcaswell would you care to chime in with your opinion on proper names?

dot-asm replied:
> What about ec_scalarmul_ladder?

Great! Except maybe ec_scalar_mul_ladder? Just pick either :-)

bbbrumley replied:
> maybe ec_scalar_mul_ladder

+1

mattcaswell replied:
`ec_scalar_mul_ladder` works for me.

romen commented:
Ok, so I should have addressed all the feedback received 'til now!

- the renaming of `ec_mul_consttime` to `ec_scalar_mul_ladder` happens early, in the very first commit after the one removing the stale SM2 error codes, as it made more sense given also the changes to the internal documentations happening in that commit (but the function remains `static`). 
- `CHANGES` entries, and following commit messages are updated accordingly.
- as it was before, only the very last commit removes the `static` attribute from `ec_scalar_mul_ladder`, and adds declaration (and documentation in the form of comments) to `ec_lcl.h`, alongside the move of checks to the `EC_POINTs_mul()` wrapper.

@dot-asm @richsalz @mattcaswell I did not remove the "WIP" label yet (again) as I would like to have feedback about the decisions I made for the very first commit:

## Remove stale SM2 error codes

Run `make update ERROR_REBUILD=-rebuild` to remove some stale error codes for SM2 (which is now using its own submodule for error codes, i.e., `SM2_*`).

I filtered out copyright bumps for other unmodified headers, and also the following diff (as both files are internal and should never be included in C++ source):

```diff
diff --git a/crypto/include/internal/sm2err.h b/crypto/include/internal/sm2err.h
index f32d26c04b..9a7e2b6275 100644
--- a/crypto/include/internal/sm2err.h
+++ b/crypto/include/internal/sm2err.h
@@ -15,6 +15,9 @@

 # ifndef OPENSSL_NO_SM2

+#  ifdef  __cplusplus
+extern "C"
+#  endif
 int ERR_load_SM2_strings(void);

 /*
diff --git a/include/internal/dsoerr.h b/include/internal/dsoerr.h
index ac6bdcd038..a54a18545e 100644
--- a/include/internal/dsoerr.h
+++ b/include/internal/dsoerr.h
@@ -15,6 +15,9 @@

 # ifndef OPENSSL_NO_DSO

+#  ifdef  __cplusplus
+extern "C"
+#  endif
 int ERR_load_DSO_strings(void);

 /*
```

romen commented on crypto/ec/ec_mult.c:
```diff
@@ -108,62 +108,83 @@ void EC_ec_pre_comp_free(EC_PRE_COMP *pre)
+    }
+
+    if (!BN_mul(cardinality, group->order, group->cofactor, ctx)) {
+        ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_BN_LIB);
```
One might argue that this case might be worthy of introducing a new "reason" code, e.g.,  `EC_R_CANNOT_COMPUTE_CARDINALITY` or something similar.

What do you think about it?

bbbrumley replied:
@romen Given that after #6648 we learned defaults for order and cofactor are not NULL but 0, I would argue a very early abort `BN_is_zero(group->order) || BN_is_zero(group->cardinality)` would be more appropriate.

In which case this particular `BN_mul` failure is pretty generic.

romen replied:
I was doing this, and deciding on the proper reason code and discovered that the order could officially be INVALID, UNDEFINED, UNKNOWN or WRONG:

```bash
~/repos/openssl >>> ack -B1 'EC_R_.*_ORDER' **/*.c
crypto/ec/ec_asn1.c
781-    if (BN_is_negative(a) || BN_is_zero(a)) {
782:        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, EC_R_INVALID_GROUP_ORDER);
--
785-    if (BN_num_bits(a) > (int)field_bits + 1) { /* Hasse bound */
786:        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, EC_R_INVALID_GROUP_ORDER);

crypto/ec/ec_check.c
54-    if (BN_is_zero(order)) {
55:        ECerr(EC_F_EC_GROUP_CHECK, EC_R_UNDEFINED_ORDER);
--
61-    if (!EC_POINT_is_at_infinity(group, point)) {
62:        ECerr(EC_F_EC_GROUP_CHECK, EC_R_INVALID_GROUP_ORDER);

crypto/ec/ec_err.c
313-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_INVALID_FORM), "invalid form"},
314:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_INVALID_GROUP_ORDER), "invalid group order"},
--
356-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNDEFINED_GENERATOR), "undefined generator"},
357:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNDEFINED_ORDER), "undefined order"},
358-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNKNOWN_GROUP), "unknown group"},
359:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNKNOWN_ORDER), "unknown order"},
--
362-    "wrong curve parameters"},
363:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_WRONG_ORDER), "wrong order"},

crypto/ec/ec_key.c
299-    if (BN_is_zero(order)) {
300:        ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_INVALID_GROUP_ORDER);
--
307-    if (!EC_POINT_is_at_infinity(eckey->group, point)) {
308:        ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_WRONG_ORDER);
--
316-        if (BN_cmp(eckey->priv_key, order) >= 0) {
317:            ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_WRONG_ORDER);

crypto/ec/ec_mult.c
835-    if (BN_is_zero(order)) {
836:        ECerr(EC_F_EC_WNAF_PRECOMPUTE_MULT, EC_R_UNKNOWN_ORDER);

crypto/ec/ecp_nistz256.c
827-    if (BN_is_zero(order)) {
828:        ECerr(EC_F_ECP_NISTZ256_MULT_PRECOMPUTE, EC_R_UNKNOWN_ORDER);
```

For consistency with `ec_mult.c` and `ecp_nistz256.c` I am going with `EC_R_UNKNOWN_ORDER` (and adding a similar one for the cofactor), but to me it looks like `EC_R_UNDEFINED_ORDER` from `EC_GROUP_check` has exactly the same semantic, and it's a bit confusing!

dot-asm commented:
> Run make update ERROR_REBUILD=-rebuild ... I filtered out copyright bumps for other unmodified headers, and also the following diff (as both files are internal and should never be included in C++ source)

As for copyright bumps, I'd say it's appropriate. As for pair of extern "C", one *could* just as well leave them. Because they are auto-generated, and the script will keep re-adding them, because it makes no difference between public and private headers. One can argue that script should be fixed, but couple of extern "C" in private headers is really no biggie, and one can just as well tolerate them(*). Let's put it this way. If it's not much trouble to re-add those extern "C", do so. And if you feel that it's rather a distraction in the context of *this* request, do nothing. Script might get fixed at later point [or not].

romen commented:
So 243fb25 adds the `group->order` and `group->cofactor` checks.

Re @bbbrumley : 
> Edit: but really seems like a separate PR to me.

I think it is appropriate to have the checks (and documentation) added where `ec_scalar_mul_ladder` is made accessible to other internal functions, so I went ahead and included it in this PR.

@dot-asm : 99ae74e adds back the `extern C`, I just tried it and everything autosquashes without problems. **The extended part of the commit message for 40d87a1a49 should be purged before the final merge.**

romen commented:
Out of WIP now.

romen commented on crypto/ec/ec2_smpl.c:
```diff
@@ -737,4 +680,275 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+     * order or cofactor set to 0.
+     */
+    if (num > 1 || BN_is_zero(group->order) || BN_is_zero(group->cofactor))
+        return ec_wNAF_mul(group, r, scalar, num, points, scalars, ctx);
```
If I understand correctly, you are suggesting to handle the (`scalar != NULL && num==1`) in `ec_wNAF_mul()` calling `ec_scalar_mul_ladder`: the problem with that is that for other implementations `ec_scalar_mul_ladder` can be slower than the regular "non-const-time" path.

Double multiplication is basically only used by ECDSA verification right now, where all the inputs are public, so a fast and insecure implementation could be used, rather than `ec_scalar_mul_ladder`.

What we really would need is refactor the `EC_METHOD` API to ditch/deprecate the unified `EC_POINTs_mul` (which makes a sense from a mathematical point of view, but is not really the way the 3 different scalar multiplication cases are used by cryptosystems) in favor of 3 different function pointers for `fixed_point_mul`, `variable_point_mul`, `double_point_mul`.
Right now we have wrappers that are forcing the 3 different operations into the unified EC_POINTs_mul path, and then in `ec_scalr_mul_ladder` we split them again.
A specific `double_point_mul` could almost half the current execution cost of the verification step, as we could use formulas (e.g., derived from Shamir's trick) that avoid doing 2 distinct scalar multiplications and then 1 addition, and share as much computation as possible.
But this must wait for the next major release development cycle!

Does this answer your comment, or did I completely misunderstood what you meant?

dot-asm replied:
> the problem with that is that for other implementations ec_scalar_mul_ladder can be slower

Fair enough. Then how about
```
if (scalar == NULL || num != 1 || order == 0 || cofactor == 0)
    return ec_wNAF_mul()
/* scalar != NULL && num == 1 is faster with ec_scalar_mul_ladder here [and now] */
...
```
Just keep in mind that this is side note, so "how about" is really just a question, it's not a suggestion. Point is not larger than order != 0 && cofactor !=0 case in ec_wNAF_mul being partial duplicate of the code below this line. [Off topic, this kind of asks for question if ec_wNAF_mul is actually suitable name.]

romen replied:
I think this change would hinder readability and mask intent: what we want to convey is that binary non-degenerate (i.e., have a `order` and `cardinality` that makes sense, given that mathematically speaking 0 is not really an option for either value) curves use `ec_scalar_mul_ladder` for the 3 cases, any other crazyness that can be fed through the `EC_POINTs_mul` API is basically undocumented and not cryptographically relevant, so we will let the default wNAF implementation handle it.

### Re: renaming `ec_wNAF_mul`

I would argue that it would make sense to have in `ec_mult.c` something like `ec_points_mul_default` working as a wrapper distinguishing the two single point multiplication cases (for which `ec_scalar_mul_ladder` is then called) from the truly multi-point multiplication case (for which `ec_wNAF_mul`) is called.
But I would leave this to a separate PR.

Even better, in a separate PR I would rather start a transition to 3 separate function pointers in `EC_METHOD` for the fixed/variable/double point multiplication, and let the `EC_POINTs_mul` wrapper do something like:

```c
/* All kind of checks on not NULL args, every point_is_compat, etc. */

if (group->meth->points_mul)
    return group->meth->points_mul(...);

degenerate = BN_is_zero(group->order) || BN_is_zero(group->cofactor);

if (scalar != NULL && num == 0)
    if (group->meth->fixed_scalar_mul)
        return group->meth->fixed_scalar_mul(...);
    else if (!degenerate)
        return ec_scalar_mul_ladder(...);

if (scalar == NULL && num == 1)
    if (group->meth->variable_scalar_mul)
        return group->meth->variable_scalar_mul(...);
    else if (!degenerate)
        return ec_scalar_mul_ladder(...);

if (scalar != NULL && num == 1 && group->meth->double_scalar_mul)
    return group->meth->double_scalar_mul(...);

/* HIC SUNT DRACONES */
return ec_wNAF_mul(...);
```
Until something similar exist, I don't think we can avoid specific `EC_POINTs_mul` implementations to repeat the checks for the 3 cases.
And the existence of the degenerate case will also probably means that most non-default implementations of the 3 scalar_mul variants will need to repeat the check for degenerate cases.

richsalz commented:
This looks good to me, but I'll not approve until/unless @mattcaswell says that his concerns have been addressed.

mattcaswell commented:
I will merge shortly

romen commented:
Thanks @mattcaswell! It might have gotten lost in the conversation, but during rebase the msg for the first commit is in need of editing!
